### PR TITLE
C++: Initial attempt at IR-based value numbering

### PR DESCRIPTION
**Pull request by dave-bartolomeo**
This is a first attempt at IR-based global value numbering. It handles most IR instructions other than `Call`, but needs aliased SSA before it can determine that two loads of the same global variable are equivalent if there are no intervening stores to escaped memory.

Still a work in progress, but should be enough to try out for other analysis that needs GVN.

### CODE REVIEW

**Comment by jbj** [comment]
Why is there a GVN library for each SSA stage? Are you planning to use GVN as part of the SSA or aliasing computation?

**Comment by dave-bartolomeo** [comment]
@jbj I'm considering using GVN as part of alias analysis. Today, alias analysis can determine that a given pointers points to offset `n` in variable `x`, but it doesn't handle the common case of two accesses where we don't know what specific location is accessed, but we do know that the two accesses are to the same location:

```
int x;
int* px = &x;
*px = 5;
return *px;
```
Here, we know that `px` points to `x`, so all of the `*px` accesses are equivalent to `x`.

```
int foo(int* q) {
  *q = 7;
  return *q;
}
```
Here, we have no idea what `q` points to, so today's alias analysis just assumes that the two `*q` accesses "may partially overlap". We'd like it to recognize that the two `*q` accesses "must exactly overlap".

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/gvn/ValueNumber.qll [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  instr.getResultType() = type and
+  instr.getElementSize() = elementSize and
+  valueNumber(instr.getLeftOperand()) = leftOperand and
+  valueNumber(instr.getRightOperand()) = rightOperand
+}
+
+private predicate unaryValueNumber(UnaryInstruction instr, FunctionIR funcIR, Opcode opcode,
+    Type type, ValueNumber operand) {
+  instr.getFunctionIR() = funcIR and
+  (not instr instanceof InheritanceConversionInstruction) and
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/gvn/ValueNumber.qll]: Is it impossible to value number `InheritanceConversionInstruction`s or is it just not particularly useful at the moment?

Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/gvn/ValueNumber.qll]: That was just an oversight. Added, including test case.

**Pull request merged**